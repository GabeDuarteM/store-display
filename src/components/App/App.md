```js
window.onload = () => {
  window.onscroll = undefined
}
;<App />
```
