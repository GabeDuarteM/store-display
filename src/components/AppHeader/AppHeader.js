import React from 'react'
import PropTypes from 'prop-types'
import injectSheet from 'react-jss'

export const styles = {
  appHeader: {
    color: '#585858',
    fontFamily: 'Montserrat',
    fontSize: 24.0,
    fontWeight: 400,
    textAlign: 'center',
  },
}

const AppHeader = ({ classes }) => (
  <div className={classes.appHeader}>STORE DISPLAY</div>
)

export const propTypes = {
  /** @ignore */
  classes: PropTypes.object.isRequired,
}

AppHeader.propTypes = propTypes

export default injectSheet(styles)(AppHeader)
