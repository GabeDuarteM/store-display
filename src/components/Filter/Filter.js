import React from 'react'
import PropTypes from 'prop-types'
import injectSheet from 'react-jss'
import cn from 'classnames'

import GenderRadioButtonFilterContainer from '../../containers/GenderRadioButtonFilterContainer'
import SortDropdownFilterContainer from '../../containers/SortDropdownFilterContainer'

export const styles = {
  filter: {
    backgroundColor: 'white',
    paddingTop: 20,
    paddingBottom: 25,
    paddingLeft: 62,
    paddingRight: 62,
    display: 'flex',
    justifyContent: 'space-between',
  },
}

const Filter = ({ classes, className }) => (
  <div className={cn(classes.filter, className)}>
    <GenderRadioButtonFilterContainer />
    <SortDropdownFilterContainer />
  </div>
)

export const propTypes = {
  /** @ignore */
  classes: PropTypes.object.isRequired,
  /** Apply a classname to the root component */
  className: PropTypes.string,
}

export const defaultProps = {
  className: null,
}

Filter.propTypes = propTypes
Filter.defaultProps = defaultProps

export default injectSheet(styles)(Filter)
