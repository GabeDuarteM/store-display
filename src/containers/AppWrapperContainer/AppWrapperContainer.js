import React from 'react'

import PropTypes from 'prop-types'

import configureStore from '../../store'
import AppWrapper from '../../components/AppWrapper'

const store = configureStore()

const AppWrapperContainer = ({ children }) => (
  <AppWrapper store={store}>{children}</AppWrapper>
)

export const propTypes = {
  /** The children that is supposed to be rendered */
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
}

AppWrapperContainer.propTypes = propTypes

export default AppWrapperContainer
