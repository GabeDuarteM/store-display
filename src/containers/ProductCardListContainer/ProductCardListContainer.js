import React from 'react'

import PropTypes from 'prop-types'
import { connect } from 'react-redux'

import ProductCardList from '../../components/ProductCardList'
import { filterLoadMoreProducts } from '../../store/reducers/filter/filterActions'
import {
  productRefreshList,
  productSetWishlistProp,
} from '../../store/reducers/product/productActions'

export class ProductCardListContainer extends React.Component {
  componentDidMount = () => {
    this.props.productRefreshList()
  }

  toggleIsWishlisted = id => {
    const wishlistProducts =
      JSON.parse(localStorage.getItem('wishlistProducts')) || []
    let newWishlistProducts
    const productInsideWishlist = wishlistProducts.find(
      wishlistId => wishlistId === id,
    )

    if (productInsideWishlist) {
      newWishlistProducts = [
        ...wishlistProducts.filter(productId => productId !== id),
      ]
    } else {
      newWishlistProducts = [...wishlistProducts, id]
    }
    localStorage.setItem(
      'wishlistProducts',
      JSON.stringify(newWishlistProducts),
    )
    this.props.productSetWishlistProp({
      id,
      isWishlisted: !productInsideWishlist,
    })
  }

  render() {
    const { handleClickLoadMore, products, isLastPage, className } = this.props

    return (
      <ProductCardList
        className={className}
        products={products}
        handleClickLoadMore={handleClickLoadMore}
        toggleIsWishlisted={this.toggleIsWishlisted}
        isLastPage={isLastPage}
      />
    )
  }
}

export const propTypes = {
  /** A className to apply to the root component  */
  className: PropTypes.string,
  /** @ignore */
  handleClickLoadMore: PropTypes.func.isRequired,
  /** @ignore */
  productRefreshList: PropTypes.func.isRequired,
  /** @ignore */
  productSetWishlistProp: PropTypes.func.isRequired,
  /** @ignore */
  products: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      imgUrl: PropTypes.string.isRequired,
      gender: PropTypes.string.isRequired,
      rating: PropTypes.number.isRequired,
      price: PropTypes.number.isRequired,
      isNew: PropTypes.bool.isRequired,
      isWishlisted: PropTypes.bool,
      name: PropTypes.string.isRequired,
      brand: PropTypes.string.isRequired,
    }),
  ),
  /** @ignore */
  isLastPage: PropTypes.bool,
}

export const defaultProps = {
  products: [],
  isLastPage: null,
  className: null,
}

ProductCardListContainer.defaultProps = defaultProps
ProductCardListContainer.propTypes = propTypes

const mapStateToProps = state => ({
  products: state.product.products,
  isLastPage: state.product.isLastPage,
})

const mapDispatchToProps = dispatch => ({
  handleClickLoadMore: () => {
    dispatch(filterLoadMoreProducts())
    dispatch(productRefreshList())
  },
  productRefreshList: () => dispatch(productRefreshList()),
  productSetWishlistProp: ({ id, isWishlisted }) =>
    dispatch(productSetWishlistProp({ id, isWishlisted })),
})

export default connect(mapStateToProps, mapDispatchToProps)(
  ProductCardListContainer,
)
