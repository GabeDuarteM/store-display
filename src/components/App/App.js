import React from 'react'

import injectSheet from 'react-jss'
import PropTypes from 'prop-types'

import AppHeader from '../AppHeader'
import Filter from '../Filter'
import ProductCardListContainer from '../../containers/ProductCardListContainer'

export const styles = {
  appHeaderSection: {
    margin: '34px 0',
  },
  filterSection: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: 'white',
    zIndex: 1,
    '&.sticky': {
      position: 'fixed',
      top: 0,
      width: '100%',
      boxShadow: '0px 1px 2px rgba(0,0,0,0.2)',
    },
  },
  filter: {
    width: 900,
  },
  productCardListSection: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
  },
  productCardList: {
    width: 930,
  },
  '@global .sticky + section': {
    marginTop: 115,
  },
}

const App = ({ classes }) => (
  <div>
    <section className={classes.appHeaderSection}>
      <AppHeader />
    </section>
    <section id="filterSection" className={classes.filterSection}>
      <Filter className={classes.filter} />
    </section>
    <section className={classes.productCardListSection}>
      <ProductCardListContainer className={classes.productCardList} />
    </section>
  </div>
)

window.onload = () => {
  const filterSection = document.getElementById('filterSection')

  const sticky = filterSection.offsetTop

  const makeFilterStickyIfNecessary = () => {
    if (window.pageYOffset >= sticky) {
      filterSection.classList.add('sticky')
    } else {
      filterSection.classList.remove('sticky')
    }
  }
  window.onscroll = makeFilterStickyIfNecessary
}

export const propTypes = {
  /** @ignore */
  classes: PropTypes.object.isRequired,
}

App.propTypes = propTypes

export default injectSheet(styles)(App)
