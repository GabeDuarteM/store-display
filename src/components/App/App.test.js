import React from 'react'

import { shallow, mount } from 'enzyme'
import toJson from 'enzyme-to-json'

import App, { styles, propTypes } from './App'
import AppHeader from '../AppHeader'
import Filter from '../Filter'
import AppWrapperContainer from '../../containers/AppWrapperContainer'
import ProductCardListContainer from '../../containers/ProductCardListContainer'

const createWrapperShallow = () => shallow(<App />)

const createWrapperMount = () =>
  mount(
    <AppWrapperContainer>
      <App />
    </AppWrapperContainer>,
  )

describe('COMPONENT: <App />', () => {
  it('should render without crashing', () => {
    createWrapperMount()
  })

  it("should match component's snapshot", () => {
    const wrapper = createWrapperShallow()
    expect(toJson(wrapper.dive())).toMatchSnapshot()
  })

  it("should match component's style snapshot", () => {
    expect(styles).toMatchSnapshot()
  })

  it("should match component's propTypes snapshot", () => {
    expect(propTypes).toMatchSnapshot()
  })

  it('should have an AppHeader component', () => {
    const wrapper = createWrapperShallow()

    const appHeaderExists = wrapper
      .dive()
      .find(AppHeader)
      .exists()
    expect(appHeaderExists).toBeTruthy()
  })

  it('should have an Filter component', () => {
    const wrapper = createWrapperShallow()

    const filterExists = wrapper
      .dive()
      .find(Filter)
      .exists()
    expect(filterExists).toBeTruthy()
  })

  it('should have a ProductCardListContainer component', () => {
    const wrapper = createWrapperShallow()

    const productCardListContainerExists = wrapper
      .dive()
      .find(ProductCardListContainer)
      .exists()
    expect(productCardListContainerExists).toBeTruthy()
  })

  it('should not have a sticky class on the filter section when the whole element is visible', () => {
    const section = document.createElement('section')
    section.setAttribute('id', 'filterSection')

    document.body.appendChild(section)
    window.onload()
    window.onscroll()

    const hasStickyClass = document
      .getElementById('filterSection')
      .classList.contains('sticky')
    expect(hasStickyClass).toBe(true)
  })
})
