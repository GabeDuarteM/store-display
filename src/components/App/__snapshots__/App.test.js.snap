// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`COMPONENT: <App /> should match component's propTypes snapshot 1`] = `
Object {
  "classes": [Function],
}
`;

exports[`COMPONENT: <App /> should match component's snapshot 1`] = `
<div>
  <section
    className="App-appHeaderSection"
  >
    <Jss(AppHeader) />
  </section>
  <section
    className="App-filterSection"
    id="filterSection"
  >
    <Jss(Filter)
      className="App-filter"
    />
  </section>
  <section
    className="App-productCardListSection"
  >
    <Connect(ProductCardListContainer)
      className="App-productCardList"
    />
  </section>
</div>
`;

exports[`COMPONENT: <App /> should match component's style snapshot 1`] = `
Object {
  "@global .sticky + section": Object {
    "marginTop": 115,
  },
  "appHeaderSection": Object {
    "margin": "34px 0",
  },
  "filter": Object {
    "width": 900,
  },
  "filterSection": Object {
    "&.sticky": Object {
      "boxShadow": "0px 1px 2px rgba(0,0,0,0.2)",
      "position": "fixed",
      "top": 0,
      "width": "100%",
    },
    "alignItems": "center",
    "backgroundColor": "white",
    "display": "flex",
    "justifyContent": "center",
    "zIndex": 1,
  },
  "productCardList": Object {
    "width": 930,
  },
  "productCardListSection": Object {
    "alignItems": "center",
    "display": "flex",
    "justifyContent": "center",
  },
}
`;
